apps/nl_means/nl_means_generator.cpp

```diff
@@ -86,51 +86,69 @@ class NonLocalMeans : public Halide::Generator<NonLocalMeans> {
 
         if (auto_schedule) {
             // nothing
-        } /*else if (get_target().has_gpu_feature()) {
-            // TODO: the GPU schedule is currently using to much shared memory
-            // because the simplifier can't simplify the expr (it can't cancel
-            // the 'x' term in min(((a + (x + b)) + c) - min(x + d + e))) so
-            // it ends up using the entire image size as the shared memory size.
+        } else if (get_target().has_gpu_feature()) {
+            // 22 ms on a 2060 RTX
+            Var xii, yii;
+
             non_local_means.compute_root()
-                .reorder(c, x, y).unroll(c)
-                .gpu_tile(x, y, xi, yi, 16, 8);
-            d.compute_at(non_local_means_sum, s_dom.x)
-                .tile(x, y, xi, yi, 2, 2)
-                .unroll(xi)
-                .unroll(yi)
-                .gpu_threads(x, y);
-            blur_d_y.compute_at(non_local_means_sum, s_dom.x)
-                .unroll(x, 2).gpu_threads(x, y);
-            blur_d.compute_at(non_local_means_sum, s_dom.x)
-                .gpu_threads(x, y);
-            non_local_means_sum.compute_at(non_local_means, x)
-                .gpu_threads(x, y)
+                .reorder(c, x, y)
+                .unroll(c)
+                .gpu_tile(x, y, xi, yi, 32, 16);
```

Comment:
Once again, a please for explanation of magic numbers in the comments. Future You will thank Present You.

Reply:
In this case there's no meaning behind these numbers. It's just a size that worked out OK performance-wise.

Reply:
...that sounds like a useful comment then.

Reply:
Isn't that basically the "default comment" on every schedule? Schedules inherently involve a ton of these constants that are mostly just found via trial and error, or simply set arbitrarily and forgotten about. Unless there's some specific reason for a constant to be something (e.g. a tile size needs to fit in shared/local memory, satisfies some other constraint), I think it's OK to not explain schedule constants.

Reply:
Probably overthinking it, but I've been internally holding these two same points of view since Steven's comment. For production code, I think a comment would be unnecessary here for the reason Dillon states. Future me is not going to be confused about tile sizes in schedules - I'll just assume that I tried a few things and this one was fast. But this is example code for people learning Halide too, and I think a comment here would add some value for those people, even if it just says "A 32x16 tile seems to work well. Found through trial and error".

Reply:
What he said. This is (at least partly) example code. Overdoing the explanation is arguably worthwhile in this case.